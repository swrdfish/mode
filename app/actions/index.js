export const addMessage = (uid, text, isMine) => ({
  type: 'ADD_MSG',
  uid,
  text,
  isMine
})

export const deleteMessage = (id) => ({
  type: 'DEL_MSG',
  id
})

export const setContactFilter = (username) => ({
  type: 'SET_CONTACT_FILTER',
  username
})

export const addUser = (userData) => ({
  type: 'ADD_USER',
  userData
})

export const removeUser = (userData) => ({
  type: 'DEL_USER',
  userData
})

export const changeUser = (userData) => ({
  type: 'CHG_USER',
  userData
})

export const login = (ref, username, uid, publicIp) => ({
  type: 'LOGIN',
  ref,
  username, 
  uid,
  publicIp
})

export const chat = (uid) => ({
  type: 'CHAT_USER',
  uid
})

export const notify = (message, type) => ({
  type: 'ADD_NOTIFICATION',
  notificationType: type,
  message
})

export const unnotify = (index) => ({
  type: 'DEL_NOTIFICATION',
  index
})

export const joinRoom = (connectionManager, roomRef) => ({
  type: 'JOIN_ROOM',
  connectionManager,
  roomRef
})