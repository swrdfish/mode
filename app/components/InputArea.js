import React from 'react'


class InputArea extends React.Component {
	render() {
		return (
			<div className="input-area-wrapper">
				<input type="text" placeholder="Type message" className="input-area-inputfiled" />
				<i className="fa fa-paperclip" id="attach-button" aria-hidden="true"></i>
			</div>
		)
	}
}

export default InputArea