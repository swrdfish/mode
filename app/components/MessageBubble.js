import React from 'react'
import AutoLinker from 'autolinker'
import escapeHTML from 'escape-html'


class MessageBubble extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        let item = this.props.item
        let className = item.isMine?'message-bubble mine':'message-bubble'
        let downloadButtonClass = item.isMine?"":"fa fa-download"
        let pretext = item.isMine?"Sent ":"Download "
        let text

        let itemRendered
        if(item.type == "file") {
            if(item.body.count > 1) {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.count} files&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            } else {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.filenames[0]}&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            }

            if(item.isMine) {
                text = "" + item.body.count
                text += item.body.count > 1?" files":" file"
            }
            className += " file"
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span>
                            {pretext}&nbsp;&nbsp;{text}               
                        </span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        } else {
            let messageBody = AutoLinker.link(escapeHTML(item.body), {className: "message-bubble-link"})
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span className='message-bubble-body' dangerouslySetInnerHTML={{ __html: messageBody }}></span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        }
        return itemRendered
    }
}

export default MessageBubble