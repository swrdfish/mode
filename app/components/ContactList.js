import React from 'react'
import { connect } from 'react-redux'
import fuzzysearch from 'fuzzysearch'

import ContactProfilePicture from './ContactProfilePicture.js'
import { chat } from '../actions'


class ContactList extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            selected: ""
        }
        this.handleClick = this.handleClick.bind(this)
    }

    handleClick(e) {
        let selected = e.currentTarget.getAttribute("data-id")
        if (selected) {
            this.setState({selected})
            this.props.dispatch(chat(selected))
            this.props.room.connectionManager.getConnection(selected)
        }
    }

    render() {  
        let userList = this.props.userList || []
        let userToDisplay = userList.map(user => ({
                "name": user.name? user.name : user.localIP,
                "id": user.uid,
                "unseen": user.unseen
            })
        )

        if(this.props.searchText !== ''){
            userToDisplay = userToDisplay.filter(contact => (fuzzysearch(this.props.searchText, contact.name.toLowerCase())))
        }

        let contacts = userToDisplay.map(contact => {
            let className = this.state.selected === contact.id ? 'contact-list-item selected' : 'contact-list-item'
            let unseen = contact.unseen?(
                <div className="unseen-count-wrapper">
                    <span className="unseen-count">{contact.unseen}</span>
                </div>
            ) : ''
            return (
                <div className={className} key={contact.id} data-id={contact.id} onClick={this.handleClick} >
                    <ContactProfilePicture profilePicture="/static/img/default_contact.jpg" />
                    <div className="contact-name">
                        {contact.name}
                    </div>
                    {unseen}
                </div>
            );
        }, this);

        return (
            <div>
                {(contacts.length > 0)? contacts : (<div className="contact-not-found">No contacts found.</div>)}
            </div>
        )
    }
}


const mapStateToProps = ({userList, contactFilter, room, dispatch}) => ({
    searchText: contactFilter,
    userList,
    room,
    dispatch
})

let ConnectedContactList = connect(mapStateToProps)(ContactList) 
export default ConnectedContactList