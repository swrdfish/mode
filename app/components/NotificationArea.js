import React from 'react'
import { connect } from 'react-redux'
import {unnotify} from '../actions'


class NotificationArea extends React.Component {
  constructor(props){
    super(props)
    this.handleClearNotification = this.handleClearNotification.bind(this)
  }

  handleClearNotification(event) {
    this.props.dispatch(unnotify(event.target.getAttribute('data-notification-idx')))
  }

  render() {
    let displayNotifications = this.props.notifications.map((notify, index) => {
      return (
        <div className={notify.type == 'error'? 'notification error': 'notification'} key={index}>
          {notify.msg}
          <i className="fa fa-close clear-notification" aria-hidden="true" data-notification-idx={index} onClick={this.handleClearNotification}></i>
        </div>
      )
    })

    return (
      <div className="notification-area">
        {displayNotifications}
      </div>
    )
  }
}

const mapStateToProps = ({notifications, dispatch}) => ({
  notifications,
  dispatch
})

let ConnectedNotificationArea = connect(mapStateToProps)(NotificationArea) 
export default ConnectedNotificationArea;