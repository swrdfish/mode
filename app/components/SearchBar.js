import React from 'react'
import { connect } from 'react-redux'

import { setContactFilter } from '../actions'


export class SearchBar extends React.Component{
	constructor(props) {
		super(props)
		this.state = {value: ''}
		this.handleOnChange = this.handleOnChange.bind(this)
		this.handleClearSearch = this.handleClearSearch.bind(this)
		this.handleKeyDown = this.handleKeyDown.bind(this)
		this.closeButton = (<i className="fa fa-close clear-search" aria-hidden="true" onClick={this.handleClearSearch}></i>)
		this.searchButton = (<i className="fa fa-search" aria-hidden="true"></i>)
	}

	handleOnChange(event) {
		this.setState({value: event.target.value})
		this.props.dispatch(setContactFilter(event.target.value.toLowerCase()))
	}

	handleClearSearch(event) {
		this.setState({value: ''})
		this.props.dispatch(setContactFilter(''))
	}

	handleKeyDown(event) {
		if(event.key === "Escape"){
			event.stopPropagation()
			event.preventDefault()
			this.setState({value: ''})
			this.props.dispatch(setContactFilter(''))
		}
	}

	render() {
		return (
			<div className="search-wrapper">
				<input
					placeholder="search users"
					value={this.state.value}
					onKeyDown={this.handleKeyDown}
					onChange={this.handleOnChange}
					type="text"
				/>
				{this.state.value != '' ? this.closeButton : this.searchButton}
			</div>
		)
	}
}


let ConnectedSearchBar = connect()(SearchBar) 
export default ConnectedSearchBar;
