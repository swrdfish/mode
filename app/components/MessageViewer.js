import React from 'react'
import MessageBubble from "./MessageBubble"


class MessageViewer extends React.Component {
    render(){
        if(this.props.chatter) {
            let renderedMsg = (<div className='empty-message-view'>Say something!</div>)
            if (this.props.messages) {
                let messages = this.props.messages.reduceRight((arr, item) => {
                    return arr.concat((<MessageBubble item={item} key={item.id} onDownload={this.props.onDownload}/>))
                }, [])
                renderedMsg = (
                    <div className='messages-scrollable'>
                        {messages}
                    </div>
                )
            }

            return (
                <div className='message-view-wrapper'>
                    {renderedMsg}
                </div>
            )
        } else {
            return (
                <div className='message-view-wrapper'>
                    <div className='empty-message-view'>select someone to chat with.</div>
                </div>
            )
        }
    }
}

export default MessageViewer