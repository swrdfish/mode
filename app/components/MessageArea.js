import React from 'react'
import { connect } from 'react-redux'

import InputArea from './InputArea.js'
import MessageViewer from './MessageViewer.js'


class MessageArea extends React.Component {
    render() {
        let connectionManager = this.props.room.connectionManager
        let chatter = this.props.chatArea.uid
        return (
            <div id="message-area" className="col-md-9">
                <MessageViewer connectionManager={connectionManager} chatter={chatter} />
                <InputArea />
            </div>
        )
    }
}

let mapStateToProps = ({chatArea, room}) => ({
  chatArea,
  room
})

const ConnectedMessageArea = connect(mapStateToProps)(MessageArea)
export default ConnectedMessageArea