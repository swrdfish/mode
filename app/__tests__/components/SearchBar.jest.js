import React from 'react'
import { shallow } from 'enzyme'
import { SearchBar } from '../../components/SearchBar'
import { setContactFilter } from '../../actions'
import renderer from 'react-test-renderer';

describe('Component: SearchBar', () => {
    it('should render the SearchBar properly', () => {
        const wrapper = shallow(<SearchBar />)
        expect(wrapper.find(".search-wrapper")).toHaveLength(1)
        expect(wrapper.find("input")).toHaveLength(1)
        expect(wrapper.find(".fa-search")).toHaveLength(1)
    })

    it('should match previous snapshot', () => {
        const tree = renderer
            .create(<SearchBar />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should handle onChange event', () => {
        const dispatch = jest.fn()
        let handleOnChange = jest.spyOn(SearchBar.prototype, 'handleOnChange')
        let wrapper = shallow(<SearchBar dispatch={dispatch} />)
        wrapper.find("input").simulate('change', {target: {value: "FOO"}})
        expect(handleOnChange).toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith(setContactFilter("foo"))
        expect(wrapper.state().value).toEqual("FOO")
        expect(wrapper.find("input").prop('value')).toEqual("FOO")
    })

    it('should show clear button', () => {
        let wrapper = shallow(<SearchBar />)
        wrapper.setState({value: "FOO"})
        expect(wrapper.find(".clear-search")).toHaveLength(1)
    })

    it('should handle escape key', () => {
        const handleKeyDown = jest.spyOn(SearchBar.prototype, 'handleKeyDown')
        const stopPropagation = jest.fn()
        const preventDefault = jest.fn()
        const dispatch = jest.fn()
        const event = {
            key: 'Escape',
            stopPropagation,
            preventDefault
        }
        let wrapper = shallow(<SearchBar dispatch={dispatch} />)
        wrapper.setState({value: "FOO"})
        wrapper.find("input").simulate('keydown', event)
        expect(handleKeyDown).toHaveBeenCalled()
        expect(stopPropagation).toHaveBeenCalled()
        expect(preventDefault).toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith(setContactFilter(''))
        expect(wrapper.state().value).toEqual('')
    })

    it('should handle escape key only', () => {
        const handleKeyDown = jest.spyOn(SearchBar.prototype, 'handleKeyDown')
        const stopPropagation = jest.fn()
        const preventDefault = jest.fn()
        const dispatch = jest.fn()
        const event = {
            key: 'h',
            stopPropagation,
            preventDefault
        }
        let wrapper = shallow(<SearchBar dispatch={dispatch} />)
        wrapper.setState({value: "FOO"})
        wrapper.find("input").simulate('keydown', event)
        expect(handleKeyDown).toHaveBeenCalled()
        expect(stopPropagation).not.toHaveBeenCalled()
        expect(preventDefault).not.toHaveBeenCalled()
        expect(dispatch).not.toHaveBeenCalledWith()
        expect(wrapper.state().value).toEqual('FOO')
    })

    it('should handle clear search', () => {
        const handleClearSearch = jest.spyOn(SearchBar.prototype, 'handleClearSearch')
        const dispatch = jest.fn()
        let wrapper = shallow(<SearchBar dispatch={dispatch} />)
        wrapper.setState({value: "FOO"})
        wrapper.find(".clear-search").simulate('click')
        expect(handleClearSearch).toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith(setContactFilter(''))
        expect(wrapper.state().value).toEqual('')
        expect(wrapper.find('.fa-search')).toHaveLength(1)  
    })
})