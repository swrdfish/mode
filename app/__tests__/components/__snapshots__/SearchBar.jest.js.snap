// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchBar should match previous snapshot 1`] = `
<div
  className="search-wrapper"
>
  <input
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="search users"
    type="text"
    value=""
  />
  <i
    aria-hidden="true"
    className="fa fa-search"
  />
</div>
`;
